feat: [FC-0047] Full-Bleed Header + Top Navigation

IvanStepanok opened the pull request:
This PR introduces updated navigation and banner for the course home.

Light and Dark theme:
![Simulator Screen Recording - iPhone 15 Pro Max - 2024-04-02 at 17 06 12](https://github.com/openedx/openedx-app-ios/assets/128456094/1f55e47d-9ead-4267-842d-0748ef0fdf7b) ![Simulator Screen Recording - iPhone 15 Pro Max - 2024-04-02 at 17 15 01](https://github.com/openedx/openedx-app-ios/assets/128456094/99166587-d409-4687-be6b-f593523ae5b1)

In the horizontal state, the top bar is always collapsed.
![Simulator Screen Recording - iPhone 15 Pro Max - 2024-04-02 at 17 19 22](https://github.com/openedx/openedx-app-ios/assets/128456094/dc40b578-65b4-4141-9acc-47cae902f81f)

The iPad version doesn't have a collapsing effect.
![Simulator Screen Recording - iPad (10th generation) - 2024-04-02 at 17 22 35](https://github.com/openedx/openedx-app-ios/assets/128456094/388d8c79-94f9-493a-b090-c154b7d74066)

:rotating_light: **Breaking changes** :rotating_light::
The `COURSE_BANNER_ENABLED` and `COURSE_TOP_TAB_BAR_ENABLED` feature flags have been removed.

openedx-webhooks commented:
Thanks for the pull request, @IvanStepanok! Please note that it may take us up to several weeks or months to complete a review and merge your PR.

Feel free to add as much of the following information to the ticket as you can:
- supporting documentation
- [Open edX discussion forum threads](https://discuss.openedx.org/)
- timeline information ("this must be merged by XX date", and why that is)
- partner information ("this is a course on edx.org")
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will be done via the GitHub pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).


This is currently a draft pull request. When it is ready for our review and all tests are green, click "Ready for Review", or remove "WIP" from the title, as appropriate.

<!-- comment:end_of_wip -->




<!-- comment:external_pr -->
<!-- data: eyJkcmFmdCI6IHRydWV9 -->

marcotuts reviewed (approved):
Product review only 👍

rnr commented:
Issues with loading indicator:
1) Changed loading indicator for Course Home view (is it approved?):

Before|After
--|--
<img width="280" src="https://github.com/openedx/openedx-app-ios/assets/37253/ae709b20-3c07-4b4d-b56f-0f5f517a15f0">|<img width="280" src="https://github.com/openedx/openedx-app-ios/assets/37253/2e974192-f138-4ec3-95a5-e924c799b913">

2) Loading indicator for "More" tab:
<img width="280" src="https://github.com/openedx/openedx-app-ios/assets/37253/7b224765-f048-4153-b5dc-b24fef2d35ec">

rnr commented:
Issues with Scroll
1) When user scrolls on Discussion tab it doesn't collapse header (how it does on Home tab for example)

https://github.com/openedx/openedx-app-ios/assets/37253/3d66d306-396e-431c-9dbf-d4d07ea53166

2) Can't scroll to bottom on Dates tab (I guess it happens when content takes up a bit more space than is available on the screen). Strange animation for this case:

https://github.com/openedx/openedx-app-ios/assets/37253/74a1b81f-6d77-4226-bfdb-264da01fbcdf

saeedbashir commented:
The progress bar location is odd if we use pull to refresh while data is loading. and If we interact with UI white loading the bleeding header isn't working properly. 
https://github.com/openedx/openedx-app-ios/assets/5606473/161d1427-be69-420f-858a-10a14341aca7

IvanStepanok commented:
## @rnr Thanks for your review!🙌
1. Old loading indicators are returned. 
![Simulator Screenshot - iPhone 15 Pro Max - 2024-04-18 at 13 44 02](https://github.com/openedx/openedx-app-ios/assets/128456094/05ac3ffc-244b-4c9d-938f-3dea674237bb)

2. Fixed the “More” loading indicator.
![Simulator Screenshot - iPhone 15 Pro Max - 2024-04-18 at 13 45 19](https://github.com/openedx/openedx-app-ios/assets/128456094/fe59241e-33b7-4063-bbdb-2e8531a4f146)

3. Pull to refresh indicators are equal now.
![Simulator Screenshot - iPhone 15 Pro Max - 2024-04-18 at 13 45 48](https://github.com/openedx/openedx-app-ios/assets/128456094/968747d8-2675-429e-8726-8fb54911e7c3) ![Simulator Screenshot - iPhone 15 Pro Max - 2024-04-18 at 13 45 42](https://github.com/openedx/openedx-app-ios/assets/128456094/17bad2df-f9c9-40e8-af98-16a785fe8a19)

## Scroll issues
1. The discussion tab collapsing issue is fixed.

https://github.com/openedx/openedx-app-ios/assets/128456094/d34c4464-da9c-4d95-b08f-df2efec52c52

2. The animation glitch on some scroll views is fixed by adding new logic.

https://github.com/openedx/openedx-app-ios/assets/128456094/7666636d-e581-4fe7-81b5-15681eca1321

If the content height in ScrollView is not enough to collapse the header, collapsing mode are disabled. Thanks for paying attention to this bug, it's very important.🙌

## @saeedbashir Thanks for your review, too!❤️

1. Fixed a bug with changes in header size after loading data.
2. The header is no longer transparent during loading.
3. Progress bars are currently in the normal position.

https://github.com/openedx/openedx-app-ios/assets/128456094/b69909f2-198f-48da-8da0-7554da32484b

If you discover any new bugs, I'd be delighted to fix them!

rnr commented:
> If the content height in ScrollView is not enough to collapse the header, collapsing mode are disabled.
I think we need to disable collapse mode for the 'More' tab as it tries to collapse, but we have two rows (Distributions and Announcements) at the moment. Thank you

https://github.com/openedx/openedx-app-ios/assets/37253/a10624a8-7030-4582-8b95-349083877d4f

rnr commented:
Is it a design decision to disable this collapse mode for iPad in general?


https://github.com/openedx/openedx-app-ios/assets/37253/b480a039-e9c7-417f-b9e1-a9514c94b27f

saeedbashir commented on Course/Course/Presentation/Subviews/TopHeaderView.swift:
```diff
@@ -0,0 +1,173 @@
+//
+//  TopHeaderView.swift
```
How about giving it a more meaningful name like course header view or some as per its functionality Thoughts?

IvanStepanok replied:
Renamed to CourseHeaderView

saeedbashir commented on Course/Course/Presentation/Subviews/TopHeaderView.swift:
```diff
@@ -0,0 +1,173 @@
+
+struct TopHeaderView: View {
+    
+    @ObservedObject var viewModel: CourseContainerViewModel
+    var title: String
+    @Binding var collapsed: Bool
+    var containerWidth: CGFloat
+    var animationNamespace: Namespace.ID
+    @Binding private var isAnimatingForTap: Bool
+    @Environment(\.isHorizontal) private var isHorizontal
+    
+    let collapsedHorizontalHeight: CGFloat = 230
+    let collapsedVerticalHeight: CGFloat = 260
+    let expandedHeight: CGFloat = 300
```
All these variables can be private.

IvanStepanok replied:
Thanks! Fixed

saeedbashir commented on Course/Course/Presentation/Subviews/TopHeaderView.swift:
```diff
@@ -0,0 +1,173 @@
+    let collapsedVerticalHeight: CGFloat = 260
+    let expandedHeight: CGFloat = 300
+    
+    private struct GeometryName {
+        static let backButton = "backButton"
+        static let topTabBar = "topTabBar"
+        static let blurSecondaryBg = "blurSecondaryBg"
+        static let blurPrimaryBg = "blurPrimaryBg"
+        static let blurBg = "blurBg"
+    }
```
Do we need a struct here, how about converting it to an enum?

IvanStepanok replied:
Thanks! Good idea

saeedbashir commented on Course/Course/Presentation/Subviews/TopHeaderView.swift:
```diff
@@ -0,0 +1,173 @@
+                        .allowsHitTesting(false)
+                        .clipped()
+                        .background(Theme.Colors.background)
+                }
+                }
+                .disabled(true)
+                .ignoresSafeArea()
```
The code needs formatting here.

IvanStepanok replied:
sure

saeedbashir commented on Course/Course/Presentation/Subviews/TopHeaderView.swift:
```diff
@@ -0,0 +1,173 @@
+        .ignoresSafeArea(edges: .top)
+    }
+    
+    private func topTabBar(containerWidth: CGFloat) -> some View {
```
How about renaming it to courseMenuBar or courseTabbar? thoughts?

saeedbashir commented on Core/Core/View/Base/RefreshProgressView.swift:
```diff
@@ -0,0 +1,22 @@
+
+import SwiftUI
+
+public struct RefreshProgressView: View {
```
I guess it's being used for pull to refresh, how about renaming it as per functionality thoughts?

saeedbashir commented on Core/Core/View/Base/ResponsiveView.swift:
```diff
@@ -0,0 +1,70 @@
+
+import SwiftUI
+
+public struct ResponsiveView: View {
```
Is it possible to give it a meaningful name as every view is by default responsive view.

IvanStepanok replied:
Renamed to DynamicOffsetView. Good?

saeedbashir commented on Core/Core/View/Base/ResponsiveView.swift:
```diff
@@ -0,0 +1,70 @@
+    private let expandedHeight: CGFloat = 240
+    private let coordinateBoundaryLower: CGFloat = -115
+    
+    @State var lastHeight: CGFloat = 240
```
It can be private

IvanStepanok replied:
Thanks! Fixed

saeedbashir commented on Core/Core/View/Base/ResponsiveView.swift:
```diff
@@ -0,0 +1,70 @@
+        .overlay(
+            GeometryReader { geometry -> Color in
+                guard idiom != .pad, enableCollapsing else {
+                    return Color.clear
```
Color is redundant, you can directly use .clear.

IvanStepanok replied:
Thanks! Fixed

saeedbashir reviewed (changes requested):
* The header is not collapsing on home and videos tab.
* The back button is not showing the menu on long press.

https://github.com/openedx/openedx-app-ios/assets/5606473/419ae958-b49a-4f7e-a447-7687c4507113

IvanStepanok commented:
https://github.com/openedx/openedx-app-ios/assets/128456094/2ed59cae-2fde-4290-adf6-3edcfd47cc53

@rnr, @saeedbashir, thanks for the review, I appreciate it. Feel free to reach out if you find any new bugs or have questions. Thanks again!

saeedbashir reviewed (dismissed):
LGTM 👍

IvanStepanok commented:
> Is it a design decision to disable this collapse mode for iPad in general?
> 
>  Screen.Recording.2024-04-19.at.11.37.37.mov

Yes, it is approved with @sdaitzman

forgotvas commented:
Header bar hides logo by half or more on iPad:


iPhone| iPad Portrait| iPad Landscape
--|--|--
![Simulator Screenshot - iPhone 15 - 2024-04-22 at 12 36 26](https://github.com/openedx/openedx-app-ios/assets/480059/b6bd7e82-0ade-436e-ac22-2b436e12698c)|![big1](https://github.com/openedx/openedx-app-ios/assets/480059/b26fc0a1-853d-4265-86de-ddc44d063200)|![big2](https://github.com/openedx/openedx-app-ios/assets/480059/72f55a12-e6d1-4309-b448-8f4930567e00)

forgotvas commented:
Back button in wrong position for pages with header on iPad:

Course home| Course home (Landscape) | Content | Content (Landscape)
--|--|--|--
![Simulator Screenshot - iPad (10th generation) - 2024-04-22 at 12 39 01](https://github.com/openedx/openedx-app-ios/assets/480059/a05e061f-5c9d-4cd7-8a0e-897022ad4221)|![Simulator Screenshot - iPad (10th generation) - 2024-04-22 at 12 39 34](https://github.com/openedx/openedx-app-ios/assets/480059/f3efc6e5-3e29-4943-8649-f07eea1fdf98)|![Simulator Screenshot - iPad (10th generation) - 2024-04-22 at 12 39 03](https://github.com/openedx/openedx-app-ios/assets/480059/c1ff21b3-bce1-4fe1-9a7d-7f0920bc97c3)|![Simulator Screenshot - iPad (10th generation) - 2024-04-22 at 12 39 31](https://github.com/openedx/openedx-app-ios/assets/480059/6d6cb023-a217-45f3-ae9c-6c2c556cc875)

forgotvas commented:
Controller with header doesn't have navigation title:

https://github.com/openedx/openedx-app-ios/assets/480059/26971cc7-bc31-4a92-b997-4b9c1c434bef

forgotvas commented:
Resume button doesn't have paddings on top:

develop | FC-0047
--|--
![Simulator Screenshot - iPad (10th generation) - 2024-04-22 at 12 46 30](https://github.com/openedx/openedx-app-ios/assets/480059/c906841c-ff4f-43cd-aa1a-96ed07bd3880)|![Simulator Screenshot - iPad (10th generation) - 2024-04-22 at 12 46 32](https://github.com/openedx/openedx-app-ios/assets/480059/ef614945-1393-4982-baf4-ae3fc10fe991)

forgotvas commented:
Pull to refresh indicator disappeared from videos tab:

Develop:
https://github.com/openedx/openedx-app-ios/assets/480059/4ec46061-f710-4ca4-9e59-31c51a35a5a9

FC-0047:
https://github.com/openedx/openedx-app-ios/assets/480059/21c9e8b9-091a-4179-beb1-022c0661eac6

forgotvas commented on Course/Course/Presentation/Outline/ContinueWithView.swift:
```diff
@@ -40,7 +40,6 @@ struct ContinueWithView: View {
                         UnitButtonView(type: .continueLesson, action: action)
                             .frame(width: 200)
                     } .padding(.horizontal, 24)
-                        .padding(.top, 32)
```
Are you sure that we don't need that?

forgotvas commented:
Thanks, @IvanStepanok, just one was skipped - back button has wrong position it should be at same position as on content page: 

course home | content
--|--
![Simulator Screenshot - iPad (10th generation) - 2024-04-22 at 22 35 47](https://github.com/openedx/openedx-app-ios/assets/480059/39d8b7da-afae-48d3-865c-1bb42951166b)|![Simulator Screenshot - iPad (10th generation) - 2024-04-22 at 22 42 11](https://github.com/openedx/openedx-app-ios/assets/480059/20c6c39d-3787-426d-afa0-65dae1dbae85)

forgotvas reviewed (approved):
LGTM
